esp-hal/src/dma/mod.rs

```diff
@@ -256,6 +250,34 @@ impl DmaDescriptor {
         next: core::ptr::null_mut(),
     };
 
+    /// Resets the descriptor for a new RX transfer.
+    fn reset_for_rx(&mut self) {
+        // Give ownership to the DMA
+        self.set_owner(Owner::Dma);
+
+        // Clear this to allow hardware to set it when the peripheral returns an EOF
+        // bit.
+        self.set_suc_eof(false);
+
+        // Clear this to allow hardware to set it when it's
+        // done receiving data for this descriptor.
+        self.set_length(0);
+    }
+
+    /// Resets the descriptor for a new TX transfer.
+    fn reset_for_tx(&mut self, is_circular: bool) {
+        // Give ownership to the DMA
+        self.set_owner(Owner::Dma);
+
+        // The `suc_eof` bit doesn't affect the transfer itself, but signals when the
+        // hardware should trigger an interrupt request. In circular mode,
+        // we set the `suc_eof` bit for every buffer we send. We use this for
+        // I2S to track progress of a transfer by checking OUTLINK_DSCR_ADDR.
+        // In non-circular mode, we only set it for the last descriptor to signal the
+        // end of the transfer.
+        self.set_suc_eof(self.next.is_null() || is_circular);
```

Comment:
Sorry I don't understand. What's `with_eof`, and why would you prefer having this logic be moved out into the 4 callers in two versions?

Reply:
> What's `with_eof`

I just mean you can pass in a boolean to directly set `set_suc_eof`, rather than have the logic in here.

> and why would you prefer having this logic be moved out into the 4 callers in two versions?

Because this isn't the only way to decide how to set the tx `eof` bit.